src/api/ApiCall.hh

```diff
@@ -0,0 +1,143 @@
+#ifndef RA_API_CALL_HH
+#define RA_API_CALL_HH
+#pragma once
+
+#include "services\IThreadPool.hh"
+#include "services\ServiceLocator.hh"
+
+#include <string>
+
+namespace ra {
+namespace api {
+
+enum class ApiResult
+{    
+    None = 0,       // unspecified
+    Success,        // call was successful
+    Error,          // an error occurred
+    Failed,         // call was unsuccessful, but didn't error
+    Unsupported,    // call is not supported and was not made
+};
+
+struct ApiRequestBase
+{
+protected:
+    ApiRequestBase() noexcept = default;
+    ~ApiRequestBase() noexcept = default;
+    ApiRequestBase(const ApiRequestBase&) = default;
+    ApiRequestBase& operator=(const ApiRequestBase&) = default;
+    ApiRequestBase(ApiRequestBase&&) noexcept = default;
+    ApiRequestBase& operator=(ApiRequestBase&&) noexcept = default;
+
+    template<class TRequest, class TCallback>
+    static void CallAsync(const TRequest& request, TCallback&& callback)
+    {
+        ra::services::ServiceLocator::GetMutable<ra::services::IThreadPool>().RunAsync([request, callback = std::move(callback)]
+            {
+                callback(request.Call());
+            });
+    }
+};
+
+struct ApiResponseBase
+{
+protected:
+    ApiResponseBase() noexcept = default;
+    ~ApiResponseBase() noexcept = default;
+    ApiResponseBase(const ApiResponseBase&) = default;
+    ApiResponseBase& operator=(const ApiResponseBase&) = default;
+    ApiResponseBase(ApiResponseBase&&) noexcept = default;
+    ApiResponseBase& operator=(ApiResponseBase&&) noexcept = default;
+
+public:
+    /// <summary>
+    /// The overall result of the API call.
+    /// </summary>
+    ApiResult Result;
+
+    /// <summary>
+    /// An error message associated with a failed call.
+    /// </summary>
+    std::string ErrorMessage;
+
+    /// <summary>
+    /// Gets whether the API call was considered successful.
+    /// </summary>
+    /// <returns><c>true</c> if the call was successful, <c>false</c> if not.</returns>
+    bool Succeeded() const noexcept
+    {
+        return (Result == ApiResult::Success);
+    }
+
+    /// <summary>
+    /// Gets whether the API call was considered unsuccessful.
+    /// </summary>
+    /// <returns><c>true</c> if the call was unsuccessful, <c>false</c> if not.</returns>
+    bool Failed() const noexcept
+    {
+        return (Result == ApiResult::Error || Result == ApiResult::Unsupported || Result == ApiResult::Failed);
+    }
+};
+
+// ---- template for generating a subclass file (replace XXXX with API name) ----
+// NOTE: Implementation of Call method will be in ApiCall.cpp to avoid references IServer and all of its dependencies
```

Comment:
The _following_ template can be used for any new APIs - just replace XXXX with the API name, and flesh out the request and response parameters. Then implement the `Call` redirect in `ApiBase` and the actual implementation in `ConnectedServer` and/or the other `IServer` implementations.